src/lua/protobuf.lua

```diff
@@ -0,0 +1,616 @@
+        error(('Input number value %f for %q is not integer'):format(
+            value, field_def.name))
+    elseif type(value) == 'cdata' and not is_number64(value) then
+        error(('Input cdata value %q for %q field is not integer'):format(
+            ffi.typeof(value), field_def.name))
+    end
+end
+
+-- }}} Global helpers
+
+
+-- {{{ is_scalar, is_enum, is_message
+
+local function is_scalar(field_def)
+    return scalars[field_def.type]
+end
+
+local function is_enum(protocol, field_def)
+    return protocol[field_def.type].type == 'enum'
+end
+
+local function is_message(protocol, field_def)
+    return protocol[field_def.type].type == 'message'
+end
+
+-- }}} is_scalar, is_enum, is_message
+
+
+-- {{{ Validation
+
+local function validate_length(value)
+    local MAX_LEN = 2^32
+    if string.len(value) > MAX_LEN then
+        error("Too long string to be encoded")
+    end
+end
+
+local function validate_table_is_array(field_def, value)
+    assert(type(value) == 'table')
+    local key_count = 0
+    local min_key = 1/0
+    local max_key = -1/0
+    for k, _ in pairs(value) do
+        if type(k) ~= 'number' then
+            error(('Input array for %q repeated field ' ..
+                'contains non-numeric key: %q'):format(field_def.name, k))
+        end
+        if k - math.floor(k) ~= 0 then
+            error(('Input array for %q repeated field contains ' ..
+                'non-integer numeric key: %q'):format(field_def.name, k))
+        end
+        key_count = key_count + 1
+        min_key = math.min(min_key, k)
+        max_key = math.max(max_key, k)
+    end
+    if key_count == 0 then
+        return
+    end
+    if min_key ~= 1 then
+        error(('Input array for %q repeated field got min index %d. ' ..
+            'Must be 1'):format(field_def.name, min_key))
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value.'):format(
+            field_def.name, field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_range(field_def, value, range)
+    local min = range ~= nil and range[1] or nil
+    local max = range ~= nil and range[2] or nil
```

Comment:
I checked this function. It works correct with number and int64_t values, but it is incorrect with uint64_t. In previous version of this function we forbid the comparison with the lower limit for uint64_t values because `10ULL < -1/0` returns `true`. In my implementation `nil` in range prevents the comparison with this limit. But it seems that I would better return comment about this pecularity.